packages/e2e-test-utils-playwright/src/admin/visit-site-editor.ts

```diff
@@ -37,31 +37,32 @@ export async function visitSiteEditor(
 		query.set( 'canvas', canvas );
 	}
 
-	const canvasLoader = this.page.locator(
-		// Spinner was used instead of the progress bar in an earlier version of
-		// the site editor.
-		'.edit-site-canvas-loader, .edit-site-canvas-spinner'
-	);
-
 	await this.visitAdminPage( 'site-editor.php', query.toString() );
 
-	// Try waiting for the canvas loader to appear first, so that the locator
-	// that waits for it to disappear doesn't resolve prematurely.
-	await canvasLoader.waitFor().catch( () => {} );
-
 	/**
 	 * @todo This is a workaround for the fact that the editor canvas is seen as
 	 * ready and visible before the loading spinner is hidden. Ideally, the
 	 * content underneath the loading overlay should be marked inert until the
 	 * loading is done.
 	 */
-	await canvasLoader.waitFor( {
-		state: 'hidden',
-		// Bigger timeout is needed for larger entities, like the Large Post
-		// HTML fixture that we load for performance tests, which often doesn't
-		// make it under the default timeout value.
-		timeout: 60_000,
-	} );
+	if ( ! query.size || postId || canvas === 'edit' ) {
```

Comment:
I think this check is probably a bit fragile and not something that will stay consistent over time. A better check would be to check if the "preview area" exist in the DOM or something like that.

Reply:
I agree, although checking if something exists is what we had before this PR, and it didn't work well. We were checking whether the loader was there, and if not, we were swallowing the `waitFor()` promise error and continuing with the test. The downside of such a solution is that whenever the loader element is not there, we add the current `actionTimeout` value to the test (before the error is swallowed), which is 10s for E2E tests and 120s for perf tests. We could explicitly fix that timeout for that `waitFor()` call to, e.g., 5s, but it would still add 5s for no-loader situations and wouldn't guarantee stability. 

The most stable solution for these situations IMO is to use `locator.or` or parallel locator waits with `Promise.race`. We need to know which element is there when the canvas is not there and vice-versa. For example:

```js
await Promise.race([
  page.locator('.canvas-loader').waitFor(),
  page.locator('.some-panel-instead-of-canvas').waitFor(),
]);
```

I couldn't quickly trace an element like that when making this fix. Do you know any we could use here?